File: Data Structures & Algorithms/house-robber/submission-5.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        prev = 0
        curr = nums[0]
        for i in range(1, len(nums)):
            take = nums[i] + prev
            notake = curr
            prev = curr
            curr = max(take, notake)
        return curr

File: Data Structures & Algorithms/jump-game/submission-5.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        vector<int> dp(nums.size(), 0);
        dp[nums.size()-1] = 1;
        for (int i = nums.size()-2; i >= 0; i--) {
            if (nums[i] == 0) continue;
            for (int j = 1; j <= nums[i]; j++) {
                if (i + j < nums.size()) {
                    dp[i] = dp[i] || dp[i+j];
                }
            }
        }
        return dp[0];
    }
};
